### ISSUE TITLE
Network visitor problem

### ISSUE THREAD

**Issue by daisyron** [davisking/dlib#933]
Hi,

I would like to get the outputs of the different DNN outputs after running it on some input.

I wrote a Visitor class:
```
class DumpNetOutputsVistor {

public:
  template <typename any_net>
  void operator()(size_t idx, any_net l) {
    const tensor& lOutput = l.get_output();
    ...
  }
  template <typename T, typename U, typename E>
  void operator()(size_t , dlib::add_layer<T,U,E>& l) {
    const tensor& lOutput = l.get_output();
    ...
  }
};
```

which get the layer output and dump it.

In my network I use a general image input class (T type image with `nofChannels` channels) that I wrote, having the following template trace:
```
template <typename T, size_t nofChannels, size_t NR, size_t NC=NR>
class input_general_image_sized
```

When I compile the code I get the error:
```
error: ‘class dlib::input_general_image_sized<float, 3ul, 90ul, 45ul>’ has no member named ‘get_output’
```

So... I have to implement the get_input method for my input image layer. 

How do I do it? As far as I understand this layer is some kind of adaptation layer that does not contain the data it self... 
What is the expected behavior of this method for an input layer?

Thanks and regards,
Ron.

**Comment by davisking** [comment]
No, input layers never implement a `get_output()`.  Something else is going wrong in your code if you are getting an error like that.

**Comment by daisyron** [comment]
Hi David,

This is my network definition:
`
enum { NOF_INPUT_CHANNELS = 3 };
enum { NOF_IN_IMAGE_ROWS = 90 };
enum { NOF_IN_IMAGE_COLS = 45 };
enum { NOF_OUT_CLASSES = 2 };

typedef input_general_image_sized<float,NOF_INPUT_CHANNELS,NOF_IN_IMAGE_ROWS,NOF_IN_IMAGE_COLS> InputTensorType;
typedef InputTensorType::input_type input_image_type;
typedef InputTensorType::pixel_type my_pixel_type;

using cnnNetType = 
      loss_multiclass_log<
      fc<NOF_OUT_CLASSES,
      relu<fc<20,
      max_pool<2,2,2,2,relu<con<32,3,3,1,1,
      max_pool<2,2,2,2,relu<con<32,3,3,1,1,
      max_pool<2,2,2,2,relu<con<64,5,5,1,1,
      max_pool<2,2,2,2,relu<con<64,7,7,1,1,InputTensorType
      >>>>>>>>>>>>>>>>;
`

This is my network definition:
`
    cnnNetType net;
    dlib::deserialize(netName) >> net;
`

This is my network forward run:
`
auto label = net(inputImage);
`

and this is my layer visiting call:
`
    DumpNetOutputsVistor visitor;
    dlib::visit_layers(net.subnet(),visitor);
`

Any problem with this?

Thanks,

Ron.

**Comment by davisking** [comment]
You need more overloads in your visitor to catch each case.  You are calling `.get_output()` in your visitor on the input layer.  You need to add an overload to the visitor that catches the input layer and does something else.

**Comment by daisyron** [comment]
How can I tell that the current layer in the visitor is an input layer?

**Comment by daisyron** [comment]
typeid ?

**Comment by davisking** [comment]
You are calling  `.get_output()` in the general case of "any_net".  Don't do that.  Then if there are other layer types you want to do something with in addition to add_layer you can add overloads for them as well.  But your code is always calling `.get_output()` on every layer.  But like you noticed, input layers don't have a `.get_output()`.

**Comment by daisyron** [comment]
Thanks for you help...

In `DumpNetOutputsVisitoe` I did:
`
class DumpNetOutputsVistor {
public:
  template <typename any_net>
  void operator()(size_t idx, any_net l) {
    const tensor& lOutput = l.get_output();
    ...
  }
  void operator()(size_t idx, InputTensorType l) { }
};
`

So for the special case of `InputTensorType` I do nothing...

Is this seems better?

Regards,
Ron.

**Comment by daisyron** [comment]
Hi,

After compilation went successfully, I have a run time issue. When I run it I get the error:
`0 fc     (num_outputs=2) learning_rate_mult=1 weight_decay_mult=1 bias_learning_rate_mult=1 bias_weight_decay_mult=0`
`1 relu`
`2 fc     (num_outputs=20) learning_rate_mult=1 weight_decay_mult=1 bias_learning_rate_mult=1 bias_weight_decay_mult=0`
`Accessing this layer's get_output() is disabled because an in-place layer has been stacked on top of it.`

So it seems that the second fc layer is an in-place layer?

What is wrong here?

Thanks,
Ron.

**Comment by davisking** [comment]
You need to fix the formatting on your posts so the code isn't impossible to read.

**Comment by daisyron** [comment]
Hi,

Here its is again... 

In the visitor implementation I added printing of each layer details. Therefore in the following you can see that the last printed layer is the second `fc` layer after which the get_output fails...

After compilation went successfully, I have a run time issue. When I run it I get the error:

0 fc (num_outputs=2) learning_rate_mult=1 weight_decay_mult=1 bias_learning_rate_mult...
1 relu
2 fc (num_outputs=20) learning_rate_mult=1 weight_decay_mult=1 bias_learning_rate_mult=1 ...
Accessing this layer's get_output() is disabled because an in-place layer has been stacked on top of it.

So it seems that the second fc layer is an in-place layer?

What is wrong here?

Thanks,
Ron.

**Comment by davisking** [comment]
relu layers overwrite the memory used by their lower layers.  This makes things run faster and use less RAM.  So there literally isn't any output buffer to see from that fc layer.

**Comment by daisyron** [comment]
Thanks for your reply...

Two comments:
1. Is this means that I don't have access to the output of relu layers, so I'll not be able to observe their output?
2. What should I do in order to prevent the program to crash? Any change to the Visitor?

Ron.

**Comment by daisyron** [comment]
Any comment please?

**Comment by davisking** [comment]
You can access the output of relu layers, it's the layer that came before the relu that is not available.

You should avoid calling get_outputs() on layers that don't have outputs available.  You need to think about what exactly you are trying to accomplish to decide how best to do that.  Like why you are trying to do this in the first place.  Why is it ok to get some outputs and not others? Things like that.

**Comment by daisyron** [comment]
I Just need to get the outputs of the different layers of a network for some input...

Is there a better way to do that than using a visitor?

**Comment by davisking** [comment]
There are a lot of ways to do it, you could call layer<10>(net) to get
layer 10 and just enumerate them.  Or you could use a visitor.  But what
are you going to do if you just get them all?  Skip the ones that happen to
have had their outputs erased?  That doesn't seem like a good idea.  There
has to be something you know about your problem that makes it ok to ignore
some outputs and pay attention to others.  That's what you have to think
about.

**Comment by daisyron** [comment]
Generally I’m interested in the outputs of all layers in the DNN. The dlib implementation prevent does not enable that due to the in-place layers implementations. So I thought to just ignore their outputs while using a visitor.

The visitor approach is generic... calling a specific layer, such as layer<10>(net), will require changing the dumping code for each network...
